contracts/contracts/sbtc-bootstrap-signers.clar

```diff
@@ -58,3 +67,79 @@
     )
 )
 
+;; Multisig generation
+
+;; Generate the p2sh redeem script for a multisig
+(define-read-only (pubkeys-to-spend-script
+    (pubkeys (list 15 (buff 65)))
+    (m uint)
+  )
+  (concat (uint-to-byte (+ u80 m)) ;; "m" in m-of-n
+  (concat (pubkeys-to-bytes pubkeys) ;; list of pubkeys with length prefix
+  (concat (uint-to-byte (+ u80 (len pubkeys))) ;; "n" in m-of-n
+  0xae ;; CHECKMULTISIG
+  )))
+)
+
+;; hash160 of the p2sh redeem script
+(define-read-only (pubkeys-to-hash
+    (pubkeys (list 15 (buff 65)))
+    (m uint)
+  )
+  (hash160 (pubkeys-to-spend-script pubkeys m))
+)
+
+;; Given a set of pubkeys and an m-of-n, generate a principal
+(define-read-only (pubkeys-to-principal
+    (pubkeys (list 15 (buff 65)))
+    (m uint)
+  )
+  (unwrap-panic (principal-construct?
+    (if is-in-mainnet 0x14 0x15) ;; address version 
+    (pubkeys-to-hash pubkeys m)
+  ))
+)
+
+;; Concat a list of pubkeys into a buffer with length prefixes
+(define-read-only (pubkeys-to-bytes (pubkeys (list 15 (buff 65))))
+  (fold concat-pubkeys-fold pubkeys 0x)
+)
+
+(define-read-only (concat-pubkeys-fold (pubkey (buff 65)) (iterator (buff 1500)))
```

Comment:
Ah yeah I had that way earlier as a "lazy" way of tracking max length. I've updated this to be the exact size, which is `(15 * 33) + 15`, because each pubkey has a length prefix (1 byte)